pkg/apis/core/types.go

```diff
@@ -1967,6 +1974,9 @@ type Handler struct {
 	// TODO: implement a realistic TCP lifecycle hook
 	// +optional
 	TCPSocket *TCPSocketAction
+	// Sleep specifies a wait time.
+	// +optional
+	Sleep *SleepAction
```

Comment:
Needs better comments about why someone one use this and what the effect is WRT signal delivery.

Reply:
Handler is used in other places.  This means that `sleep` is now an option for livenessProbes, for readinessProbes, for postStart hooks, etc.  I did a very cursory analysis below of how this is PROBABLY safe for preStop, but I doubt we can (or want to) make the same guarantees about every user of Handler.

Aside from that, can we make even a hypothetical case for why sleep would make sense in any of those?

Maybe better to make a new struct, embed `Handler`, and add `sleep`; or just clone all of `Handler` into `preStopHandler`. And then fix all the related code.

Reply:
The only hypothetical use I could see for sleep in the other cases would be to build test cases / build reproduce cases.

Reply:
Agree - we should fix this to only affect preStop

Reply:
I tried copying the whole struct but then noticed that I would need to duplicate some code in pkg/kubelet/lifecycle/handlers.go. I'll go with embedding path to avoid that.

Reply:
Hm, I just tried that, but can this work with existing clusters? Wouldn't the new definition break existing clusters?

Looks like I should go back to copying the struct.

Reply:
Not sure what you mean breaking clusters, wrt Handler.

https://play.golang.org/p/u-JoOYkQC9o

This *could*, however break clusters with version skew.  Validation today demands that exactly one Handler be defined, and a kubelet could be back-rev vs an apiserver.

This is a general problem and has come up repeatedly.  Maybe we should do a release where we remove that validation pattern wherever we can find it, and then these changes would be less awful.

Otherwise we force a 2-release window on ever such change.

@liggitt @bgrant0607